dev/NavigationView/NavigationView.cpp

```diff
@@ -524,29 +596,50 @@ void NavigationView::UpdateItemsRepeaterItemsSource(const winrt::ItemsRepeater&
 
 void NavigationView::OnNavigationViewItemIsSelectedPropertyChanged(const winrt::DependencyObject& sender, const winrt::DependencyProperty& args)
 {
-    if (auto const nvib = sender.try_as<winrt::NavigationViewItemBase>())
+    if (auto const nvi = sender.try_as<winrt::NavigationViewItem>())
     {
         // Check whether the container that triggered this call back is the selected container
-        bool isContainerSelectedInModel = IsContainerTheSelectedItemInTheSelectionModel(nvib);
-        bool isSelectedInContainer = nvib.IsSelected();
+        bool isContainerSelectedInModel = IsContainerTheSelectedItemInTheSelectionModel(nvi);
+        bool isSelectedInContainer = nvi.IsSelected();
 
         if (isSelectedInContainer && !isContainerSelectedInModel)
         {
-            auto indexPath = GetIndexPathForContainer(nvib);
-            m_selectionModel.SelectAt(indexPath);
+            auto indexPath = GetIndexPathForContainer(nvi);
+            UpdateSelectionModelSelection(indexPath);
         }
         else if (!isSelectedInContainer && isContainerSelectedInModel)
         {
-            auto indexPath = GetIndexPathForContainer(nvib);
+            auto indexPath = GetIndexPathForContainer(nvi);
             auto indexPathFromModel = m_selectionModel.SelectedIndex();
 
-            if (indexPathFromModel &&
-                indexPath.GetSize() > 0 && indexPathFromModel.GetSize() > 0 &&
-                indexPath.GetAt(0) == indexPathFromModel.GetAt(0))
+            if (indexPathFromModel && indexPath.CompareTo(indexPathFromModel) == 0)
             {
                 m_selectionModel.DeselectAt(indexPath);
             }
         }
+
+        if (isSelectedInContainer)
+        {
+            nvi.IsChildSelected(false);
+        }
+    }
+}
+
+void NavigationView::OnNavigationViewItemExpandedPropertyChanged(const winrt::DependencyObject& sender, const winrt::DependencyProperty& args)
+{
+    if (auto const nvi = sender.try_as<winrt::NavigationViewItem>())
+    {
+        if (nvi.IsExpanded())
+        {
+            RaiseExpandingEvent(nvi);
+        }
+
+        ShowHideChildrenItemsRepeater(nvi);
+
+        if (!nvi.IsExpanded())
+        {
+            RaiseCollapsedEvent(nvi);
+        }
```

Comment:
Can you explain this? I don't get why we have expanding but no expanded and collapsed but no collapsing

Reply:
The user has an option to set the 'HasUnrealizedChildren' property on a NVI. This allows them to populate the children of NVI when the expanding event gets raised, which is why we need the expanding event raise before anything gets done.

---
In reply to: [386578450](https://github.com/microsoft/microsoft-ui-xaml/pull/2004#discussion_r386578450) [](ancestors = 386578450)

Reply:
We could add all 4 events, but that felt unnecessary for now. If a customer asks, I'm certainly open to adding the other two.

The idea behind only adding the two events was to provide users the right opportunity to react to user input. Expanding will fire before the item expands and Collapsed will fire after it finishes collapsing. I believe those to be the appropriate times for an app to, for example, inject child items and delete them again dynamically. (The app scenario is showing a list of test reports as children, where the tests are run on some regular schedule and the user should see the most current set.)